typegate/src/runtimes/random.ts

```diff
@@ -78,48 +78,93 @@ export class RandomRuntime extends Runtime {
+): any {
+  const config = typ.config ?? {};
+  if (Object.prototype.hasOwnProperty.call(config, "gen")) {
+    const { gen, ...arg } = config;
+    return chance[gen as string](arg);
+  }
+  switch (typ.type) {
+    case "object":
+      return {};
+    case "optional": {
+      const childNodeName = tgTypes[typ.item];
+      return chance.bool()
+        ? randomizeRecursively(childNodeName, chance, tgTypes)
+        : null;
     }
-    switch (typ.type) {
-      case "object":
-        return {};
-      case "optional": {
-        const childNodeName = this.getTgTypeNameByIndex(typ.item);
-        return this.chance.bool()
-          ? this.randomizeRecursively(childNodeName)
-          : null;
+    case "integer":
+      return chance.integer();
+    case "string":
+      if (typ.format === "uuid") {
+        return chance.guid();
+      }
+      if (typ.format === "email") {
+        return chance.email();
+      }
+      if (typ.format === "uri") {
+        return chance.url();
+      }
+      if (typ.format === "hostname") {
+        return chance.domain();
       }
-      case "integer":
-        return this.chance.integer();
-      case "string":
-        if (typ.format === "uuid") {
-          return this.chance.guid();
-        }
-        if (typ.format === "email") {
-          return this.chance.email();
-        }
-        return this.chance.string();
-      case "boolean":
-        return this.chance.bool();
-      case "list": {
-        const res = [];
-        let size = this.chance.integer({ min: 1, max: 10 });
-        const childNodeName = this.getTgTypeNameByIndex(typ.items);
-        while (size--) {
-          res.push(this.randomizeRecursively(childNodeName));
-        }
-        return res;
+      if (typ.format === "date-time") {
+        return chance.date().toISOString();
       }
-      default:
-        throw new Error(`type not supported "${typ.type}"`);
+      if (typ.format == "ean") {
+        return generateEAN(chance);
+      }
+      return chance.string();
+    case "boolean":
+      return chance.bool();
+    case "list": {
+      const res = [];
+      let size = chance.integer({ min: 1, max: 10 });
+      const childNodeName = tgTypes[typ.items];
+      while (size--) {
+        res.push(
+          randomizeRecursively(childNodeName, chance, tgTypes),
+        );
+      }
+      return res;
     }
+
+    default:
+      throw new Error(`type not supported "${typ.type}"`);
+  }
+}
+
+function generateEAN(chance: typeof Chance) {
+  let ean = "0";
+
+  for (let i = 1; i <= 11; i++) {
+    ean += chance.integer({ min: 0, max: 9 }).toString();
+  }
+
+  const checkDigit = calculateCheckDigit(ean);
+  ean += checkDigit;
+
+  return ean;
+}
+
+function calculateCheckDigit(ean: string) {
+  const digits = ean.split("").map(Number);
+
+  let sum = 0;
+  for (let i = 0; i < digits.length; i++) {
+    sum += (i % 2 === 0) ? digits[i] : digits[i] * 3;
   }
+
+  const checkDigit = (10 - (sum % 10)) % 10;
```

Comment:
shouldn't `10 - (sum % 10)` be enough? since `0 <= anything_positive % 10 <= 9`

Reply:
Done

Reply:
Actually `10 - (sum % 10)` is between `1` and `10` inclusive...

Reply:
yes indeed, my bad, that's on me.